Convert intern-cli to TypeScript

nicknisi opened the pull request:
Fixes #7 
* Convert scripts
* get all tests passing
* add gitignore
* manually use cli and ensure that commands still work
* add build script to package.json
* use dojo-loader for tests
* upgrade to latest intern for internal tests

nicknisi committed:
Convert intern-cli to TypeScript

* Convert scripts
* get all tests passing
* add gitignore
* manually use cli and ensure that commands still work
* add build script to package.json
* use dojo-loader for tests
* upgrade to latest intern for internal tests

devpaul commented on tests/intern.ts:
```diff
@@ -0,0 +1,25 @@
+export const functionalSuites: string[] = [
+];
+
+export const excludeInstrumentation = /^(?:tests|node_modules)\//;
```
This regexp should be 

```
/^(?:_build\/tests|node_modules)\//;
```

devpaul commented on package.json:
```diff
@@ -24,15 +24,23 @@
   "preferGlobal": true,
   "bin": {
-    "intern": "./bin/intern.js"
+    "intern": "./_build/src/bin/intern.js"
```
Is this handled correctly in the release script? IIRC, `release.js` copies package.json over to `_build` and releases from there. That's what DigDug does at least, however it's not using a fancy release script. We should make sure we're on the same page w/ how Intern releases.

devpaul commented on package.json:
```diff
@@ -24,15 +24,23 @@
+    "test": "node node_modules/intern/client.js config=_build/tests/intern"
   },
   "engines": {
     "node": ">0.12"
```
DigDug supports node version 4+. We can increase this.

devpaul commented on package.json:
```diff
@@ -24,15 +24,23 @@
+  },
+  "scripts": {
+    "build": "tsc",
+    "test": "node node_modules/intern/client.js config=_build/tests/intern"
```
It may not mater right now since we're not building/testing on Travis, but it'd be nice to have a linting command-line.

devpaul commented on src/bin/intern.ts:
```diff
@@ -1,19 +1,24 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import * as cli from '../lib/cli';
+import program = require('commander'); // importing like this so helpInformation can be modified below
```
This could be rewritten this way:

```ts
import { Command } from 'commander';
const program = new Command();
```

This would allow assigning `helpInformation` later on without working around esm's import rules

devpaul commented on tsconfig.json:
```diff
@@ -0,0 +1,23 @@
+		"module": "umd",
+		"moduleResolution": "node",
+		"noImplicitAny": true,
+		"noImplicitThis": true,
```
These are supported as of TS 2+ and will be going away in tslint

```
"noUnusedParameters": true,
"noUnusedLocals": true,
```

nicknisi replied:
Added. Thanks for the info!

devpaul commented on src/bin/intern.ts:
```diff
@@ -66,31 +71,31 @@ if (!cli.acceptVersion(internPackage.version, MIN_VERSION)) {
-	var cmdName = this._name;
+	let cmdName = this._name;
 	if (this._alias) {
 		cmdName = cmdName + '|' + this._alias;
```
In general we should favor using es6 template strings over string concatenation. It's easier to see where a string begins and ends and where variables are added.

devpaul commented on src/bin/intern.ts:
```diff
@@ -66,31 +71,31 @@ if (!cli.acceptVersion(internPackage.version, MIN_VERSION)) {
 
-	var cmds = [];
-	var commandHelp = this.commandHelp();
+	let cmds: string[] = [];
```
This can be a `const` and line 95 can `cmds.push(commandHelp)`

devpaul commented on src/bin/intern.ts:
```diff
@@ -120,13 +125,14 @@ program
-			if (program.commands[i].name() === commandName) {
-				command = program.commands[i];
+	.on('help', function (args: any[]) {
+		let commandName = args[0];
```
This can be a `const`

devpaul commented on src/bin/intern.ts:
```diff
@@ -259,38 +265,38 @@ program
 
-		var options = arguments[arguments.length - 1];
-		var config = options.config || path.join(TESTS_DIR, 'intern.js');
+		const options = arguments[arguments.length - 1];
```
These should come from a typed rest parameter

```ts
.action(function(... args: string[]) {
    const options: string = args[args.length - 1];
    const internCmd: string[] = args.slice(0, -1);
});
```

devpaul commented on src/bin/intern.ts:
```diff
@@ -405,20 +410,20 @@ program
-		var options = arguments[arguments.length - 1];
-		var config = options.config || path.join(TESTS_DIR, 'intern.js');
-		var internCmd = path.join(internDir, 'runner');
+		const options = arguments[arguments.length - 1];
```
These should come from a typed rest parameter (as above)

devpaul commented on src/bin/intern.ts:
```diff
@@ -1,19 +1,24 @@
+import * as resolve from 'resolve';
+import { spawn } from 'child_process';
+import opn = require('opn');
+const pkg = require('../../../package.json');
```
Make sure this works for release, especially if we don't release from the project's root directory and instead release from another location like `_build/src`

devpaul commented on src/lib/cli.ts:
```diff
@@ -44,101 +42,105 @@ exports.copy = function (src, dst) {
 	if (verbose) {
-		var util = require('util');
 		return function () {
 			process.stderr.write('>> ' + util.format.apply(util, arguments) + '\n');
```
This would read better w/ rest parameter:

```ts
return function(...args: any[]) {
    process.stderr.write(`>> ${ util.format(... args) }\n`);
};

nicknisi replied:
Doing this throws an error in TS. Please see [#4130](https://github.com/Microsoft/TypeScript/issues/4130).

devpaul commented on src/lib/cli.ts:
```diff
@@ -162,9 +164,9 @@ function getFormatter(width, prefix) {
+	const formatter = getFormatter(80, '  ');
 	format = formatter;
-	return formatter.apply(null, arguments);
+	return formatter.apply(null, args);
```
use spread here:

```ts
    return formatter(... args);
```

devpaul commented on tests/unit/all.ts:
```diff
@@ -0,0 +1,98 @@
+registerSuite({
+	name: 'lib/cli',
+
+	acceptVersion: function () {
```
use es6 functional naming:

```ts
registerSuite({
    acceptVersion() {
        // ...
    }
});
```

jason0x43 committed:
Prepare for TS conversion

jason0x43 committed:
Infrastructure updates

- Switch to @theintern/dev for build, lint, test
- Add prettier config
- Add npm 5 package-lock.json

jason0x43 committed:
Move intern 3 init files, add intern 4 stub

jason0x43 committed:
Code style updates, prettification

jason0x43 committed:
Remove jshint comments

jason0x43 committed:
Remove some unnecessary help message code

jason0x43 committed:
Add clean script

jason0x43 committed:
WIP Add a separate tests tsconfig, use strict options
